File: src/deliveryservice/internal/delivery.go
package internal

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"log"
	"log/slog"
	"math"
	"math/rand"
	"slices"
	"strconv"
	"sync"
	"time"

	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
	"google.golang.org/protobuf/types/known/emptypb"
	"google.golang.org/protobuf/types/known/timestamppb"

	pb "github.com/pongsathonn/ihavefood/src/deliveryservice/genproto"
	amqp "github.com/rabbitmq/amqp091-go"
)

type DeliveryService struct {
	pb.UnimplementedDeliveryServiceServer

	// acceptc is a map of channels used to notify the
	// waitForRiderAccept function when a rider has accepted
	// an order. The key is the orderID, which uniquely
	// identifies each channel, and the value is the
	// channel for the riderID, indicating which rider has
	// accepted the order.
	acceptc map[string]chan *pb.ConfirmRiderAcceptRequest

	// accepte is a map of channels used to communicate
	// any errors encountered during the execution of
	// waitForRiderAccept. The key is the orderID, and
	// the value is a channel for error responses.
	accepte map[string]chan error

	rabbitmq RabbitMQ
	storage  DeliveryStorage
	mu       sync.Mutex
}

func NewDeliveryService(rabbitmq RabbitMQ, storage DeliveryStorage) *DeliveryService {
	return &DeliveryService{
		rabbitmq: rabbitmq,
		storage:  storage,

		acceptc: make(map[string]chan *pb.ConfirmRiderAcceptRequest),
		accepte: make(map[string]chan error),
	}
}

// GetOrderTracking response current rider location and
// timestamp every 1 minute. if delivery status is "DELIVRED".
// It will use deliverTime for timestamp and stop rider location.
func (x *DeliveryService) GetOrderTracking(in *pb.GetOrderTrackingRequest, stream pb.DeliveryService_GetOrderTrackingServer) error {

	var timestamp time.Time

	ticker := time.NewTicker(time.Minute)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			delivery, err := x.storage.Delivery(context.TODO(), in.OrderId)
			if err != nil {
				slog.Error("failed to retrives delivery", "err", err)
				return status.Errorf(codes.Internal, "failed to retrives delivery: %v", err)
			}

			timestamp = time.Now()

			if delivery.Status == DELIVERED {
				timestamp = delivery.Timestamp.DeliverTime
			}

			err = stream.Send(&pb.GetOrderTrackingResponse{
				OrderId: delivery.OrderID,
				RiderId: delivery.RiderID,
				RiderLocation: &pb.Point{
					Latitude:  delivery.RiderLocation.Latitude,
					Longitude: delivery.RiderLocation.Longitude,
				},

				UpdateTime: timestamppb.New(timestamp),
			})
			if err != nil {
				return status.Errorf(codes.Internal, "failed to response stream: %v", err)
			}

			// stop response if order delivered
			if delivery.Status == DELIVERED {
				break
			}
		}
	}

	return nil

}

func (x *DeliveryService) GetDeliveryFee(ctx context.Context, in *pb.GetDeliveryFeeRequest) (*pb.GetDeliveryFeeResponse, error) {

	//TODO validate input

	//TODO get restaurant address , user address

	restaurantPoint := &pb.Point{Latitude: in.RestaurantLat, Longitude: in.RestaurantLong}
	userPoint := &pb.Point{Latitude: in.UserLat, Longitude: in.UserLong}

	deliveryFee, err := calculateDeliveryFee(restaurantPoint, userPoint)
	if err != nil {
		slog.Error("calculate delivery fee", "err", err)
		return nil, status.Error(codes.Internal, "failed to calculate delivery fee")
	}

	return &pb.GetDeliveryFeeResponse{DeliveryFee: deliveryFee}, nil
}

// ConfirmRiderAccept handles requests indicating that a rider has accepted an order.
// It saves the delivery order to the database. Each order should only be accepted once.
func (x *DeliveryService) ConfirmRiderAccept(ctx context.Context, in *pb.ConfirmRiderAcceptRequest) (*pb.PickupInfo, error) {

	if in.OrderId == "" || in.RiderId == "" {
		return nil, status.Error(codes.InvalidArgument, "order number and rider id must be provided")
	}

	delivery, err := x.storage.Delivery(ctx, in.OrderId)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "failed to retrive order delivery %v", err)
	}

	if delivery.Status == DELIVERED {
		return nil, status.Error(409, "order has already been accepted")
	}

	timeOut, cancel := context.WithTimeout(ctx, time.Second*5)
	defer cancel()

	select {
	// Notify that the rider has accepted the order
	case x.acceptc[in.OrderId] <- &pb.ConfirmRiderAcceptRequest{
		RiderId: in.RiderId,
		OrderId: in.OrderId,
	}:
	case <-timeOut.Done():
		return nil, status.Error(codes.Internal, "failed to notify server that order accepted")
	}

	return &pb.PickupInfo{
		PickupCode: delivery.PickupCode,
		PickupLocation: &pb.Point{
			Latitude:  delivery.PickupLocation.Latitude,
			Longitude: delivery.PickupLocation.Longitude,
		},
		Destination: &pb.Point{
			Latitude:  delivery.Destination.Latitude,
			Longitude: delivery.Destination.Longitude,
		},
	}, nil

}

func (x *DeliveryService) ConfirmOrderDeliver(ctx context.Context, in *pb.ConfirmOrderDeliverRequest) (*emptypb.Empty, error) {

	if in.OrderId == "" {
		return nil, status.Error(codes.InvalidArgument, "order number must be provided")
	}

	if _, err := x.storage.UpdateStatus(ctx, in.OrderId, DELIVERED); err != nil {
		return nil, status.Errorf(codes.Internal, "failed to update delivery status %v", err)
	}

	return &emptypb.Empty{}, nil
}

// -------------------------------------------------------------------------------------------------------

func (x *DeliveryService) StartConsume() {

	deliveries, err := x.rabbitmq.Subscribe(
		context.TODO(),
		"rider.assign.queue", // queue
		"order.placed.event", // routing key
	)

	if err != nil {
		slog.Error("subscribe failed", "err", err)
	}

	for msg := range deliveries {
		x.deliveryAssignment() <- msg
	}

	select {}
}

// DeliveryAssignment handles incoming orders  and saves the placed order
// to the database, finds the nearest riders, and notifies them. It waits
// for rider acceptance and responds with order pickup details if the order
// is accepted.
//
// When saving the order with SaveOrderDelivery, the order status remains
// "unaccepted" until a rider accepts the order.
//
// messages is channel for receive new orders.
func (x *DeliveryService) deliveryAssignment() chan<- amqp.Delivery {

	messages := make(chan amqp.Delivery)

	for msg := range messages {

		var placeOrder pb.PlaceOrder

		if err := json.Unmarshal(msg.Body, &placeOrder); err != nil {
			slog.Error("unmarshal failed", "err", err)
			continue
		}

		go func(order *pb.PlaceOrder) {

			if order == nil {
				slog.Error("place order is empty")
				return
			}

			riders, pickup, err := x.prepareOrderDelivery(order)
			if err != nil {
				slog.Error("prepare order delivery", "err", err)
				return
			}

			if err := notifyToRider(riders, pickup); err != nil {
				slog.Error("notify to riders", "err", err)
				return
			}

			if err := x.waitForRiderAccept(order.OrderId, riders, pickup); err != nil {
				slog.Error("waiting rider accept", "err", err)
				return
			}

			err = x.rabbitmq.Publish(context.TODO(),
				"rider.assigned.event",
				amqp.Publishing{
					Type: "string",
					Body: []byte(order.OrderId),
				},
			)
			if err != nil {
				slog.Error("publish event", "err", err)
				return
			}

		}(&placeOrder)
	}

	return messages

}

// prepareOrderDelivery will  calculate the nearest riders and generate order pickup
// infomations and save new placeorder to database,
func (x *DeliveryService) prepareOrderDelivery(order *pb.PlaceOrder) ([]*pb.Rider, *pb.PickupInfo, error) {

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	riders, err := calculateNearestRider(order.UserAddress, order.RestaurantAddress)
	if err != nil {
		return nil, nil, err
	}

	pickup, err := x.generateOrderPickUp(order)
	if err != nil {
		return nil, nil, err
	}

	now := time.Now()

	orderNO, err := x.storage.Create(ctx, &newDelivery{
		OrderID:    order.OrderId,
		PickupCode: pickup.PickupCode,
		PickupLocation: &dbPoint{
			Latitude:  pickup.PickupLocation.Latitude,
			Longitude: pickup.PickupLocation.Longitude,
		},
		Destination: &dbPoint{
			Latitude:  pickup.Destination.Latitude,
			Longitude: pickup.Destination.Longitude,
		},
		Status: UNACCEPT,
		Timestamp: &dbTimeStamp{
			CreateTime:  now,
			AcceptTime:  time.Time{},
			DeliverTime: time.Time{},
		},
	})
	if err != nil {
		return nil, nil, err
	}

	slog.Info("create new order", "NO", orderNO)

	return riders, pickup, nil
}

// notifyToRider will notify to all nearest riders
//
// TODO implement push notification
func notifyToRider(riders []*pb.Rider, pickup *pb.PickupInfo) error {

	var riderIDs []string
	for _, rider := range riders {
		riderIDs = append(riderIDs, rider.Id)
	}

	notifyInfo := struct {
		riderIDs   []string
		pickupCode string
	}{
		riderIDs:   riderIDs,
		pickupCode: pickup.PickupCode,
	}

	// Example log notify
	log.Printf("[NOTIFY INFO]: %+v\n", notifyInfo)

	return nil
}

// waitRiderAccept waits for a rider to accept the order.
// The function listens for a rider's acceptance from the riderAcceptedCh channel.
// Upon receiving an acceptance:
//   - It validates if the rider was notified.
//   - Cancels the context to stop further notifications to other riders.
//   - Sends the order pickup information to the orderPickupCh channel.
//
// If no rider accepts the order within 15 minutes, the function cancels the context and stops waiting.
func (x *DeliveryService) waitForRiderAccept(orderNO string, riders []*pb.Rider,
	pickup *pb.PickupInfo) error {

	var riderIDs []string
	for _, rider := range riders {
		riderIDs = append(riderIDs, rider.Id)
	}

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	x.acceptc[orderNO] = make(chan *pb.ConfirmRiderAcceptRequest)

	for {
		select {
		case v := <-x.acceptc[orderNO]:
			if v.OrderId == "" || v.RiderId == "" {
				x.accepte[orderNO] <- errors.New("riderID or orderNO is empty")
				continue
			}

			if !slices.Contains(riderIDs, v.RiderId) {
				x.accepte[orderNO] <- fmt.Errorf("riderID %s was not notified", v.RiderId)
				continue
			}

			// Save order delivery after rider accepted
			if _, err := x.storage.UpdateRiderAccept(ctx, v.OrderId, v.RiderId); err != nil {
				slog.Error("update delivery", "err", err)
				continue
			}

			slog.Info("Rider has accepted the order",
				"orderID", v.OrderId,
				"riderID", v.RiderId,
			)

			select {
			case x.accepte[orderNO] <- nil:
			case <-ctx.Done():
				return ctx.Err()
			}

		case <-ctx.Done():
			return ctx.Err()
		}
	}

}

// generateOrderPickUp is a function that generate pickup code and locations
// to riderthat not accept order yet.
func (x *DeliveryService) generateOrderPickUp(order *pb.PlaceOrder) (*pb.PickupInfo, error) {

	code := randomThreeDigits()

	startPoint, err := x.addressToPoint(order.RestaurantAddress)
	if err != nil {
		return nil, err
	}

	destinationPoint, err := x.addressToPoint(order.UserAddress)
	if err != nil {
		return nil, err
	}

	return &pb.PickupInfo{
		PickupCode:     code,
		PickupLocation: startPoint,
		Destination:    destinationPoint,
	}, nil

}

// addressToPoint convert Address to Locations point. this function not implement
// actual Geocoding ( Google APIs ) yet . just response with example data
//
// TODO implememnt Geocoding ( Google APIs )
func (x *DeliveryService) addressToPoint(addr *pb.Address) (*pb.Point, error) {

	// [ Chaing Mai district ]
	var example = map[string]*pb.Point{
		"Mueang":    &pb.Point{Latitude: 18.7883, Longitude: 98.9853},
		"Hang Dong": &pb.Point{Latitude: 18.6870, Longitude: 98.8897},
		"San Sai":   &pb.Point{Latitude: 18.8578, Longitude: 99.0631},
		"Mae Rim":   &pb.Point{Latitude: 18.8998, Longitude: 98.9311},
		"Doi Saket": &pb.Point{Latitude: 18.8482, Longitude: 99.1403},
	}

	point, ok := example[addr.District]
	if !ok {
		return nil, fmt.Errorf("district %s invalid", addr.District)
	}

	return point, nil
}

// calculateNearestRider calculates and returns a list of riders who are
// geographically closest to the user's location, within a certain radius.
//
// This function uses the user's address to determine the proximity of available
// riders based on their current location. The algorithm should take into account
// the distance between the user's address and the riders' locations and return
// a list of the nearest riders.
func calculateNearestRider(userAddr *pb.Address, restaurantAddr *pb.Address) ([]*pb.Rider, error) {

	// TODO:
	//   - Implement the logic to calculate the radius between the user's address and
	//     the riders' locations.
	//   - Use an actual distance calculation algorithm (e.g., Haversine formula or
	//     another geo-location method) to filter riders within the radius.
	//   - Return a list of riders that are closest to the user's location.

	// riders is example data for nearest riders
	return riders, nil
}

// randomThreeDigits generate 3 digits pickup code between 100 - 999 .
func randomThreeDigits() string {

	// Half-open interval
	// - [a,b) is include a, exclude b
	// - (a,b] is include b, exclude a
	//
	// rand.Intn(900) returns a number between 0 and 899.
	// Adding 100 shifts the range to [100, 999].
	r := rand.New(rand.NewSource(time.Now().UnixNano()))
	n := r.Intn(900) + 100
	return strconv.Itoa(n)
}

// calculateDeliveryFee calculate distance from user's address to restaurant's address
//
// TODO
// Calculates the delivery fee based on the distance between the restaurant and the user,
// using the road network for routing instead of a straight-line (point A to B) distance.
func calculateDeliveryFee(userPoint *pb.Point, restaurantPoint *pb.Point) (int32, error) {

	// distance in kilometers
	distance := haversineDistance(userPoint, restaurantPoint)
	if distance < 0 || distance > 25 {
		return 0, errors.New("distance must be between 0 and 25 km")
	}

	var deliveryFee int32

	switch {
	case distance <= 5:
		deliveryFee = 0
	case distance <= 10:
		deliveryFee = 50
	default:
		deliveryFee = 100
	}

	return deliveryFee, nil
}

// haversineDistance calculates the distance between two geographic points in kilometers.
func haversineDistance(p1, p2 *pb.Point) float64 {
	const earthRadius = 6371 // Earth's radius in kilometers.

	// Convert latitude and longitude from degrees to radians.
	lat1 := p1.Latitude * math.Pi / 180
	lon1 := p1.Longitude * math.Pi / 180
	lat2 := p2.Latitude * math.Pi / 180
	lon2 := p2.Longitude * math.Pi / 180

	// Calculate the distance using the Haversine formula.
	latSin := math.Sin(lat1) * math.Sin(lat2)
	latCos := math.Cos(lat1) * math.Cos(lat2) * math.Cos(lon2-lon1)
	distance := math.Acos(latSin+latCos) * earthRadius

	return distance
}

File: src/deliveryservice/src/delivery.rs
use crate::broker::RabbitMQ;
use crate::database::Db;
use crate::ihavefood::*;
use crate::models::*;

use anyhow::{anyhow, bail, ensure, Result};
use futures::StreamExt;
use lapin::options::BasicAckOptions;
use log::error;
use prost::Message;
use rand::prelude::*;
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::Semaphore;

#[derive(Debug, Clone)]
pub struct MyDelivery {
    pub db: Arc<Db>,
    pub broker: Arc<RabbitMQ>,
    pub task_limiter: Arc<Semaphore>,
}

impl MyDelivery {
    pub async fn start_services(&self) -> Result<()> {
        tokio::try_join!(self.delivery_assignment(), self.other_event_handler())?;
        Ok(())
    }

    async fn other_event_handler(&self) -> Result<()> {
        Ok(())
    }

    async fn delivery_assignment(&self) -> Result<()> {
        let mut consumer = match self
            .broker
            .subscribe("rider.assign.queue", "order.placed.event")
            .await
        {
            Ok(consumer) => consumer,
            Err(e) => {
                error!("Failed to subscribe to delivery queue: {}", e);
                return Ok(());
            }
        };

        while let Some(delivery) = consumer.next().await {
            let db = Arc::clone(&self.db);
            let task_limiter = Arc::clone(&self.task_limiter);

            tokio::spawn(async move {
                let _permit = match task_limiter.acquire().await {
                    Ok(permit) => permit,
                    Err(e) => {
                        error!("Failed to acquire task permit: {}", e);
                        return;
                    }
                };

                if let Ok(delivery) = delivery {
                    if let Err(err) = delivery.ack(BasicAckOptions::default()).await {
                        error!("Failed to ack:{}", err);
                        return;
                    }

                    // TODO: check delivery id duplicated

                    let place_order = match PlaceOrder::decode(delivery.data.as_ref()) {
                        Ok(p) => p,
                        Err(e) => {
                            error!("Failed to decode place order:{}", e);
                            return;
                        }
                    };

                    let (riders, pickup_info) = match Self::prepare_order_delivery(&place_order) {
                        Ok(v) => v,
                        Err(e) => {
                            error!("Failed to prepare order:{}", e);
                            return;
                        }
                    };

                    let pickup_location = match pickup_info.pickup_location {
                        Some(location) => location,
                        None => {
                            error!("Empty pickup information");
                            return;
                        }
                    };

                    let drop_off_location = match pickup_info.drop_off_location {
                        Some(location) => location,
                        None => {
                            error!("Empty pickup information");
                            return;
                        }
                    };

                    if let Err(e) = db
                        .create_delivery(&NewDelivery {
                            order_id: place_order.order_id,
                            pickup_code: pickup_info.pickup_code,
                            pickup_location: DbPoint {
                                latitude: pickup_location.latitude,
                                longitude: pickup_location.longitude,
                            },
                            drop_off_location: DbPoint {
                                latitude: drop_off_location.latitude,
                                longitude: drop_off_location.longitude,
                            },
                            create_time: chrono::Utc::now(),
                        })
                        .await
                    {
                        error!("Failed to create delivery record: {}", e);
                        return;
                    }

                    Self::notify_riders(riders);
                }
            });
        }

        Ok(())
    }

    async fn handle_rider_ack(&self, order_id: String, rider_id: String) -> Result<PickupInfo> {
        self.db
            .update_delivery_rider(&order_id, &rider_id)
            .await
            .unwrap();

        let delivery = self.db.get_delivery(&order_id).await.unwrap();

        self.broker
            .publish("rider.assigned.event", order_id.as_bytes())
            .await
            .unwrap();

        Ok(PickupInfo {
            pickup_code: delivery.pickup_code,
            pickup_location: Some(Point {
                latitude: delivery.pickup_location.latitude,
                longitude: delivery.pickup_location.longitude,
            }),
            drop_off_location: Some(Point {
                latitude: delivery.drop_off_location.latitude,
                longitude: delivery.drop_off_location.longitude,
            }),
        })
    }

    pub fn prepare_order_delivery(order: &PlaceOrder) -> Result<(Vec<Rider>, PickupInfo)> {
        let riders = Self::calc_nearest_riders();
        let pickup_info = Self::generate_order_pickup(order)?;
        Ok((riders, pickup_info))
    }

    // TODO: implement some notification
    fn notify_riders(_riders: Vec<Rider>) {
        log::info!("notify to riders");
    }

    pub fn calc_delivery_fee(user_p: &Point, restau_p: &Point) -> Result<i32> {
        //distance(kilometers)
        let distance = haversine_distance(user_p, restau_p);

        ensure!(
            (0.0..=25.0).contains(&distance),
            "distance must be between 0km and 25km"
        );

        let delivery_fee: i32 = match distance {
            d if d <= 5.0 => 0,
            d if d <= 10.0 => 50,
            _ => 100,
        };

        Ok(delivery_fee)
    }

    // TODO: use actual calculation format
    fn calc_nearest_riders() -> Vec<Rider> {
        let mut riders: Vec<Rider> = Vec::new();
        (0..5).for_each(|_| riders.push(Rider::default()));
        riders
    }

    fn generate_order_pickup(order: &PlaceOrder) -> Result<PickupInfo> {
        let pickup_code = match (100..1000).collect::<Vec<i32>>().choose(&mut rand::rng()) {
            Some(code) => code.to_string(),
            None => {
                bail!("Failed to generate pickup code");
            }
        };

        let pickup_location = address_to_point(
            order
                .restaurant_address
                .as_ref()
                .ok_or_else(|| anyhow!("Restaurant address is empty"))?,
        );

        let drop_off_location = address_to_point(
            order
                .user_address
                .as_ref()
                .ok_or_else(|| anyhow!("User address is empty"))?,
        );

        Ok(PickupInfo {
            pickup_code,
            pickup_location,
            drop_off_location,
        })
    }
}

// convert Address to Point.
//
// TODO: implememnt Geocoding ( Google APIs )
fn address_to_point(address: &Address) -> Option<Point> {
    // [ Chaing Mai district ]
    let example: HashMap<&str, Point> = HashMap::from([
        (
            "Mueang",
            Point {
                latitude: 18.7883,
                longitude: 98.9853,
            },
        ),
        (
            "Hang Dong",
            Point {
                latitude: 18.6870,
                longitude: 98.8897,
            },
        ),
        (
            "San Sai",
            Point {
                latitude: 18.8578,
                longitude: 99.0631,
            },
        ),
        (
            "Mae Rim",
            Point {
                latitude: 18.8998,
                longitude: 98.9311,
            },
        ),
        (
            "Doi Saket",
            Point {
                latitude: 18.8482,
                longitude: 99.1403,
            },
        ),
    ]);

    example.get(address.district.as_str()).cloned()
}

// haversineDistance calculates the distance between two geographic points in kilometers.
fn haversine_distance(p1: &Point, p2: &Point) -> f64 {
    // Earth's radius in kilometers.
    const EARTH_RADIUS: f64 = 6371.0;

    let lat1 = p1.latitude.to_radians();
    let lon1 = p1.longitude.to_radians();
    let lat2 = p2.latitude.to_radians();
    let lon2 = p2.longitude.to_radians();

    let dlat = lat2 - lat1;
    let dlon = lon2 - lon1;

    // Haversine formula
    let a = (dlat / 2.0).sin().powi(2) + lat1.cos() * lat2.cos() * (dlon / 2.0).sin().powi(2);
    let c = 2.0 * a.sqrt().asin();

    EARTH_RADIUS * c
}

#[cfg(test)]
mod tests {
    use super::*;
    // use crate::models::*;
    use serde_json;
    use std::{fs::File, io::BufReader};

    #[test]
    fn test_calc_delivery_fee_success() {
        let user_point = Point {
            latitude: 18.7883,
            longitude: 98.9853,
        };
        let restaurant_point = Point {
            latitude: 18.6870,
            longitude: 98.8897,
        };

        let delivery_fee = MyDelivery::calc_delivery_fee(&user_point, &restaurant_point).unwrap();
        println!("{}", delivery_fee);
        assert_eq!(delivery_fee, 50);
    }

    #[test]
    fn test_calc_delivery_fee_failure() {
        let user_point = Point {
            latitude: 18.7883,
            longitude: 98.9853,
        };
        let restaurant_point = Point {
            latitude: 50.0000,
            longitude: 50.0000,
        };

        let result = MyDelivery::calc_delivery_fee(&user_point, &restaurant_point);

        assert!(result.is_err());
        assert!(result
            .unwrap_err()
            .to_string()
            .contains("distance must be between 0km and 25km"));
    }

    #[test]
    fn test_prepare_order_delivery_success() {
        // PlaceOrder needs serde deserialize (derive)
        let place_orders: Vec<PlaceOrder> = serde_json::from_reader(BufReader::new(
            File::open("/testdata/place_order_success.json").unwrap(),
        ))
        .unwrap();

        println!("placeorder ja = {:?}", place_orders);

        for place_order in place_orders {
            let result = MyDelivery::prepare_order_delivery(&place_order);

            assert!(result.is_ok());
            let (riders, pickup_info) = result.unwrap();
            assert_eq!(riders.len(), 5);
            assert!(pickup_info.pickup_location.is_some());
            assert!(pickup_info.drop_off_location.is_some());
        }
    }

    // #[test]
    // fn test_prepare_order_delivery_failure() {
    //     let order = PlaceOrder {
    //         order_id: "test_order_456".to_string(),
    //         restaurant_address: None, // Missing address
    //         user_address: Some(Address {
    //             district: "Hang Dong".to_string(),
    //             street: "User Street".to_string(),
    //             building: "User Building".to_string(),
    //         }),
    //     };
    //
    //     let result = MyDelivery::prepare_order_delivery(&order);
    //
    //     assert!(result.is_err());
    //     assert!(result
    //         .unwrap_err()
    //         .to_string()
    //         .contains("Restaurant address is empty"));
    // }
    //
    // #[test]
    // fn test_address_to_point_success() {
    //     let address = Address {
    //         district: "Mueang".to_string(),
    //         street: "Test Street".to_string(),
    //         building: "Test Building".to_string(),
    //     };
    //
    //     let result = address_to_point(&address);
    //
    //     assert!(result.is_some());
    //     let point = result.unwrap();
    //     assert_eq!(point.latitude, 18.7883);
    //     assert_eq!(point.longitude, 98.9853);
    // }
    //
    // #[test]
    // fn test_address_to_point_failure() {
    //     let address = Address {
    //         district: "Unknown District".to_string(), // Not in the HashMap
    //         street: "Test Street".to_string(),
    //         building: "Test Building".to_string(),
    //     };
    //
    //     let result = address_to_point(&address);
    //
    //     assert!(result.is_none());
    // }
    //
    // #[test]
    // fn test_haversine_distance_success() {
    //     let p1 = Point {
    //         latitude: 18.7883,
    //         longitude: 98.9853,
    //     };
    //     let p2 = Point {
    //         latitude: 18.6870,
    //         longitude: 98.8897,
    //     };
    //
    //     let distance = haversine_distance(&p1, &p2);
    //
    //     assert!(distance > 0.0);
    //     assert!(distance < 25.0); // Should be reasonable distance
    //     assert!((distance - 12.0).abs() < 5.0); // Roughly 12km +/- 5km tolerance
    // }
    //
    // #[test]
    // fn test_haversine_distance_failure() {
    //     let p1 = Point {
    //         latitude: 0.0,
    //         longitude: 0.0,
    //     };
    //     let p2 = Point {
    //         latitude: 0.0,
    //         longitude: 0.0,
    //     };
    //
    //     let distance = haversine_distance(&p1, &p2);
    //
    //     // Same point should return 0 distance
    //     assert_eq!(distance, 0.0);
    // }
    //
    // #[test]
    // fn test_calc_nearest_riders_success() {
    //     let riders = MyDelivery::calc_nearest_riders();
    //
    //     assert_eq!(riders.len(), 5);
    //     // All riders should be default instances
    //     for rider in riders {
    //         assert_eq!(rider, Rider::default());
    //     }
    // }
    //
    // #[test]
    // fn test_calc_nearest_riders_failure() {
    //     // This function always returns 5 riders, so we test the constraint
    //     let riders = MyDelivery::calc_nearest_riders();
    //
    //     // Failure case: should not return empty or wrong count
    //     assert_ne!(riders.len(), 0);
    //     assert_ne!(riders.len(), 10); // Not the expected 5
    // }
    //
    // #[test]
    // fn test_notify_riders_success() {
    //     let riders = vec![Rider::default(), Rider::default()];
    //
    //     // This function just logs, so we test it doesn't panic
    //     MyDelivery::notify_riders(riders);
    //
    //     // If we reach here, it succeeded
    //     assert!(true);
    // }
    //
    // #[test]
    // fn test_notify_riders_failure() {
    //     let riders = vec![]; // Empty riders list
    //
    //     // Should still not panic with empty list
    //     MyDelivery::notify_riders(riders);
    //
    //     // Test passes if no panic occurs
    //     assert!(true);
    // }
}
